File: lib/improved.ml
(** improved.ml — ENGINE 2: Improved LK' Prover. *)

open Types
open Formula
open Unification
open Indexing

(* ================================================================ *)
(* METAVARIABLE HANDLING                                             *)
(* ================================================================ *)

exception Timeout_exn
let metavar_counter = ref 0
let deadline = ref 0.
let fresh_metavar () = 
  let n = !metavar_counter in
  incr metavar_counter;
  "?" ^ string_of_int n

let rec get_mvars_term acc = function
  | Var x when is_metavar x -> x :: acc
  | Var _ -> acc
  | Func (_, args) -> List.fold_left get_mvars_term acc args

let rec get_mvars_formula acc = function
  | Top | Bot -> acc
  | Pred (_, args) -> List.fold_left get_mvars_term acc args
  | Not f -> get_mvars_formula acc f
  | And (f1, f2) | Or (f1, f2) | Implies (f1, f2) | Iff (f1, f2) ->
      get_mvars_formula (get_mvars_formula acc f1) f2
  | Forall (_, f) | Exists (_, f) -> get_mvars_formula acc f

let sequent_metavars s =
  let m1 = List.fold_left get_mvars_formula [] s.antecedent in
  let m2 = List.fold_left get_mvars_formula m1 s.succedent in
  unique m2

let rename_sequent_metavars s mvars =
  let mapping = List.map (fun v -> (v, Var (fresh_metavar ()))) mvars in
  let s' = apply_to_sequent mapping s in
  s', mapping

(* ================================================================ *)
(* FORMULA COMPLEXITY                                               *)
(* ================================================================ *)

let rec term_complexity = function
  | Var _ -> 1
  | Func (_, args) -> 1 + List.fold_left (fun acc t -> acc + term_complexity t) 0 args

let rec formula_complexity = function
  | Top | Bot -> 1
  | Pred (_, args) -> 1 + List.fold_left (fun acc t -> acc + term_complexity t) 0 args
  | Not f -> 1 + formula_complexity f
  | And (f1, f2) | Or (f1, f2) | Implies (f1, f2) | Iff (f1, f2) ->
      1 + formula_complexity f1 + formula_complexity f2
  | Forall (_, f) | Exists (_, f) -> 10 + formula_complexity f

(* ================================================================ *)
(* LAZY SEARCH STRUCTURE                                            *)
(* ================================================================ *)

type search_node = {
  sequent : sequent;
  expansion : (rule * search_node list) list Lazy.t;
}

(* ================================================================ *)
(* AXIOM CHECKING WITH INDEXING                                     *)
(* ================================================================ *)

let find_identity stats sigma antecedent succedent =
  let index = FingerprintIndex.create () in
  let non_atom_ants = ref [] in
  List.iter (function
    | Pred _ as p -> FingerprintIndex.insert index p
    | f -> non_atom_ants := f :: !non_atom_ants) antecedent;
  
  let rec check_succs = function
    | [] -> None
    | (Pred _ as s) :: rest ->
        let candidates = FingerprintIndex.query index s in
        let rec try_unify = function
          | [] -> check_succs rest
          | (Pred (_, args_a)) :: as_rest ->
              let args_s = match s with Pred (_, a) -> a | _ -> [] in
              (match unify ~use_rank_check:true is_metavar sigma (List.combine args_a args_s) with
               | Some sigma' -> 
                   if sigma' <> sigma then stats.inst_hits <- stats.inst_hits + 1;
                   Some sigma'
               | None -> try_unify as_rest)
          | _ :: as_rest -> try_unify as_rest
        in try_unify candidates
    | s :: rest ->
        if List.exists (formula_equal s) !non_atom_ants then Some sigma
        else check_succs rest
  in check_succs succedent

(* ================================================================ *)
(* RULE APPLICATION                                                 *)
(* ================================================================ *)

let rec collect_rules i = function
  | [] -> []
  | f :: fs ->
      let rest = collect_rules (i+1) fs in
      match f with
      | And _ -> (AndLeft i, f) :: rest
      | Or _ -> (OrLeft i, f) :: rest
      | Implies _ -> (ImpliesLeft i, f) :: rest
      | Iff _ -> (IffLeft i, f) :: rest
      | Not _ -> (NotLeft i, f) :: rest
      | Exists (x, _) -> (ExistsLeft (i, x), f) :: rest
      | Forall (x, _) -> (ForallLeft (i, Var x), f) :: rest
      | _ -> rest

let rec collect_right_rules i = function
  | [] -> []
  | f :: fs ->
      let rest = collect_right_rules (i+1) fs in
      match f with
      | And _ -> (AndRight i, f) :: rest
      | Or _ -> (OrRight i, f) :: rest
      | Implies _ -> (ImpliesRight i, f) :: rest
      | Iff _ -> (IffRight i, f) :: rest
      | Not _ -> (NotRight i, f) :: rest
      | Forall (x, _) -> (ForallRight (i, x), f) :: rest
      | Exists (x, _) -> (ExistsRight (i, Var x), f) :: rest
      | _ -> rest

let apply_rule sequent = function
  | AndLeft i ->
      let a, b = match List.nth sequent.antecedent i with And(a,b) -> a,b | _ -> failwith "AndL" in
      [{ sequent with antecedent = a :: b :: (List.filteri (fun j _ -> i <> j) sequent.antecedent) }]
  | OrRight i ->
      let a, b = match List.nth sequent.succedent i with Or(a,b) -> a,b | _ -> failwith "OrR" in
      [{ sequent with succedent = a :: b :: (List.filteri (fun j _ -> i <> j) sequent.succedent) }]
  | ImpliesRight i ->
      let a, b = match List.nth sequent.succedent i with Implies(a,b) -> a,b | _ -> failwith "ImpR" in
      [{ antecedent = a :: sequent.antecedent; succedent = b :: (List.filteri (fun j _ -> i <> j) sequent.succedent) }]
  | NotLeft i ->
      let a = match List.nth sequent.antecedent i with Not a -> a | _ -> failwith "NotL" in
      [{ antecedent = List.filteri (fun j _ -> i <> j) sequent.antecedent; succedent = a :: sequent.succedent }]
  | NotRight i ->
      let a = match List.nth sequent.succedent i with Not a -> a | _ -> failwith "NotR" in
      [{ antecedent = a :: sequent.antecedent; succedent = List.filteri (fun j _ -> i <> j) sequent.succedent }]
  | IffLeft i ->
      let a, b = match List.nth sequent.antecedent i with Iff(a,b) -> a,b | _ -> failwith "IffL" in
      let rest = List.filteri (fun j _ -> i <> j) sequent.antecedent in
      [{ sequent with antecedent = Implies(a,b) :: Implies(b,a) :: rest }]
  | ForallRight (i, _) ->
      let x, body = match List.nth sequent.succedent i with Forall(x, b) -> x, b | _ -> failwith "ForallR" in
      let avoid = List.concat_map free_vars (sequent.antecedent @ sequent.succedent) in
      let m_count = !metavar_counter in
      let x' = variant x avoid in
      let ev_name = Printf.sprintf "ev_%d_%s" m_count x' in
      [{ sequent with succedent = subst [(x, Var ev_name)] body :: (List.filteri (fun j _ -> i <> j) sequent.succedent) }]
  | ExistsLeft (i, _) ->
      let x, body = match List.nth sequent.antecedent i with Exists(x, b) -> x, b | _ -> failwith "ExistsL" in
      let avoid = List.concat_map free_vars (sequent.antecedent @ sequent.succedent) in
      let m_count = !metavar_counter in
      let x' = variant x avoid in
      let ev_name = Printf.sprintf "ev_%d_%s" m_count x' in
      [{ sequent with antecedent = subst [(x, Var ev_name)] body :: (List.filteri (fun j _ -> i <> j) sequent.antecedent) }]
  (* Branching *)
  | AndRight i ->
      let a, b = match List.nth sequent.succedent i with And(a,b) -> a,b | _ -> failwith "AndR" in
      let rest = List.filteri (fun j _ -> i <> j) sequent.succedent in
      [{ sequent with succedent = a :: rest }; { sequent with succedent = b :: rest }]
  | OrLeft i ->
      let a, b = match List.nth sequent.antecedent i with Or(a,b) -> a,b | _ -> failwith "OrL" in
      let rest = List.filteri (fun j _ -> i <> j) sequent.antecedent in
      [{ sequent with antecedent = a :: rest }; { sequent with antecedent = b :: rest }]
  | ImpliesLeft i ->
      let a, b = match List.nth sequent.antecedent i with Implies(a,b) -> a,b | _ -> failwith "ImpL" in
      let rest_ant = List.filteri (fun j _ -> i <> j) sequent.antecedent in
      [{ antecedent = rest_ant; succedent = a :: sequent.succedent };
       { antecedent = b :: rest_ant; succedent = sequent.succedent }]
  | IffRight i ->
      let a, b = match List.nth sequent.succedent i with Iff(a,b) -> a,b | _ -> failwith "IffR" in
      let rest = List.filteri (fun j _ -> i <> j) sequent.succedent in
      [{ sequent with succedent = Implies(a,b) :: rest }; { sequent with succedent = Implies(b,a) :: rest }]
  (* Instantiation *)
  | ForallLeft (i, _) ->
      let x, body = match List.nth sequent.antecedent i with Forall(x,b) -> x,b | _ -> failwith "ForallL" in
      let m = fresh_metavar () in
      [{ sequent with antecedent = subst [(x, Var m)] body :: (List.filteri (fun j _ -> i <> j) sequent.antecedent @ [Forall(x, body)]) }]
  | ExistsRight (i, _) ->
      let x, body = match List.nth sequent.succedent i with Exists(x,b) -> x,b | _ -> failwith "ExistsR" in
      let m = fresh_metavar () in
      [{ sequent with succedent = subst [(x, Var m)] body :: (List.filteri (fun j _ -> i <> j) sequent.succedent @ [Exists(x, body)]) }]
  | _ -> []

(* ================================================================ *)
(* TREE GENERATION                                                  *)
(* ================================================================ *)

let get_rule_tier = function
  | AndLeft _ | OrRight _ | ImpliesRight _ | NotLeft _ | NotRight _ | IffLeft _ -> 1 (* Invertible *)
  | ForallRight _ | ExistsLeft _ -> 2 (* Eigenvariable *)
  | AndRight _ | OrLeft _ | ImpliesLeft _ | IffRight _ -> 3 (* Branching *)
  | ForallLeft _ | ExistsRight _ -> 4 (* Instantiation *)
  | _ -> 5

let rec mk_node sequent =
  { sequent; expansion = lazy (expand_node sequent) }

and expand_node sequent =
  let l_rules = collect_rules 0 sequent.antecedent in
  let r_rules = collect_right_rules 0 sequent.succedent in
  let all_rules = List.map (fun (r, f) -> 
    (r, f, get_rule_tier r, formula_complexity f)
  ) (l_rules @ r_rules) in
  
  let sorted_rules = List.sort (fun (_, _, t1, c1) (_, _, t2, c2) ->
    if t1 <> t2 then compare t1 t2 else compare c1 c2
  ) all_rules in
  
  match sorted_rules with
  | [] -> []
  | (r, _, tier, _) :: _ ->
      if tier <= 2 then
        [(r, List.map mk_node (apply_rule sequent r))]
      else
        List.map (fun (r, _, _, _) -> (r, List.map mk_node (apply_rule sequent r))) sorted_rules

(* ================================================================ *)
(* PROOF SEARCH CORE                                                 *)
(* ================================================================ *)

let canonical_hash sigma s =
  let s' = apply_to_sequent sigma s in
  let ant = List.sort compare s'.antecedent in
  let suc = List.sort compare s'.succedent in
  Hashtbl.hash (ant, suc)

let rec search (stats : prover_stats) proof_limit inst_limit sigma failed_cache current_depth node =
  stats.steps <- stats.steps + 1;
  stats.max_depth <- max current_depth stats.max_depth;
  if stats.steps mod 1000 = 0 && Unix.gettimeofday () > !deadline then None else

  if proof_limit < 0 || inst_limit < 0 then None else
  let h = canonical_hash sigma node.sequent in
  if Hashtbl.mem failed_cache (h, proof_limit, inst_limit) then None else

  (* 1. Axiom Checks *)
  if List.mem Top node.sequent.succedent || List.mem Bot node.sequent.antecedent then
    Some sigma
  else
    match find_identity stats sigma node.sequent.antecedent node.sequent.succedent with
    | Some sigma' -> Some sigma'
    | None ->
        if proof_limit = 0 then (Hashtbl.add failed_cache (h, proof_limit, inst_limit) (); None) else
        
        (* 2. Expansion *)
        let rec try_expansions = function
          | [] -> (Hashtbl.add failed_cache (h, proof_limit, inst_limit) (); None)
          | (rule, branches) :: rest ->
              let next_inst = match rule with ForallLeft _ | ExistsRight _ -> inst_limit - 1 | _ -> inst_limit in
              if next_inst < inst_limit then stats.inst_attempts <- stats.inst_attempts + 1;
              
              if List.length branches > 1 then
                (* Branch-copying (Fix 1) *)
                let mvars = sequent_metavars node.sequent in
                let renamed_branches = List.map (fun b ->
                  let s', map = rename_sequent_metavars b.sequent mvars in
                  { b with sequent = s' }, map
                ) branches in
                
                let rec solve_and_unify current_sigma acc_results = function
                  | [] ->
                      let rec unify_origins s_final = function
                        | [] -> Some s_final
                        | m :: ms ->
                            let values = List.map (fun (sigma_i, map) ->
                              apply sigma_i (List.assoc m map)
                            ) acc_results in
                            (match values with
                             | [] | [_] -> unify_origins s_final ms
                             | v1 :: rest_vs ->
                                 let rec unify_list s acc = function
                                   | [] -> Some s
                                   | v :: vs ->
                                       (match unify ~use_rank_check:true is_metavar s [(acc, v)] with
                                        | Some s' -> unify_list s' v vs
                                        | None -> None)
                                 in
                                 (match unify_list s_final v1 rest_vs with
                                  | Some s' -> unify_origins s' ms
                                  | None -> None))
                      in
                      let merged_sigma = List.fold_left (fun acc (s_i, _) -> compose acc s_i) current_sigma acc_results in
                      unify_origins merged_sigma mvars
                  | (b, map) :: bs ->
                      (match search stats (proof_limit - 1) next_inst current_sigma failed_cache (current_depth + 1) b with
                       | Some sigma_i -> solve_and_unify current_sigma ((sigma_i, map) :: acc_results) bs
                       | None -> None)
                in
                (match solve_and_unify sigma [] renamed_branches with
                 | Some sigma_final -> Some sigma_final
                 | None -> try_expansions rest)
              else
                let rec try_branches current_sigma = function
                  | [] -> Some current_sigma
                  | b :: bs ->
                      (match search stats (proof_limit - 1) next_inst current_sigma failed_cache (current_depth + 1) b with
                       | Some sigma' -> try_branches sigma' bs
                       | None -> None)
                in
                (match try_branches sigma branches with
                 | Some sigma_final -> Some sigma_final
                 | None -> try_expansions rest)
        in
        try_expansions (Lazy.force node.expansion)

(* ================================================================ *)
(* MAIN ENTRY POINT                                                  *)
(* ================================================================ *)

let prove formula timeout_ms =
  metavar_counter := 0;
  let start_time = Unix.gettimeofday () in
  deadline := start_time +. timeout_ms /. 1000.;
  let stats = make_stats () in
  let root = mk_node { antecedent = []; succedent = [formula] } in
  let failed_cache = Hashtbl.create 1024 in
  let secs = max 1 (int_of_float (ceil (timeout_ms /. 1000.))) in
  let old_handler = Sys.signal Sys.sigalrm
    (Sys.Signal_handle (fun _ -> raise Timeout_exn)) in
  let _ = Unix.alarm secs in
  let result =
    try
      let rec iterative_deepening inst_limit =
        let elapsed = (Unix.gettimeofday () -. start_time) *. 1000. in
        if elapsed > timeout_ms || inst_limit > 10 then
          { engine = Improved; solved = false; stats = { stats with time_ms = elapsed } }
        else
          match search stats 500 inst_limit [] failed_cache 0 root with
          | Some _ ->
              stats.time_ms <- (Unix.gettimeofday () -. start_time) *. 1000.;
              { engine = Improved; solved = true; stats }
          | None -> iterative_deepening (inst_limit + 1)
      in
      iterative_deepening 0
    with Timeout_exn ->
      let elapsed = (Unix.gettimeofday () -. start_time) *. 1000. in
      { engine = Improved; solved = false; stats = { stats with time_ms = elapsed } }
  in
  let _ = Unix.alarm 0 in
  Sys.set_signal Sys.sigalrm old_handler;
  result

File: python/improved.py
"""
improved.py — ENGINE 2: Improved LK' Prover.
Ported from improved.ml.
"""

import time
from typing import Optional, Set, List, Tuple
from fol_types import (
    Term, Variable, Constant, Function,
    Formula, Atom, Negation, BinaryOp, Connective,
    QuantifiedFormula, Quantifier,
    Sequent, DerivationTree
)
from formula_utils import free_vars, variant, formula_equal, formula_complexity
from unification import unify, apply, apply_to_formula
from indexing import FingerprintIndex

class ImprovedProver:
    def __init__(self, timeout_ms: float = 50000.0):
        self.timeout_ms = timeout_ms
        self.start_time = 0.0
        self.steps = 0
        self.max_depth_reached = 0

    def _check_timeout(self) -> bool:
        elapsed = (time.time() - self.start_time) * 1000.0
        return elapsed > self.timeout_ms

    def _is_metavar(self, name: str) -> bool:
        return name.startswith("?")

    def _fresh_metavar(self, n: int) -> Tuple[str, int]:
        return f"?{n}", n + 1

    def _formula_list_equal(self, l1: List[Formula], l2: List[Formula]) -> bool:
        if len(l1) != len(l2): return False
        for f1 in l1:
            if not any(f1 == f2 for f2 in l2): return False
        return True

    def _sequent_equal(self, s1: Sequent, s2: Sequent) -> bool:
        return self._formula_list_equal(s1.antecedent, s2.antecedent) and \
               self._formula_list_equal(s1.succedent, s2.succedent)

    def _find_identity(self, sigma: dict, antecedent: List[Formula], succedent: List[Formula]) -> Optional[dict]:
        index = FingerprintIndex()
        non_atom_ants = set()
        for a in antecedent:
            if isinstance(a, Atom):
                index.insert(a)
            else:
                non_atom_ants.add(a)
        
        for s in succedent:
            if not isinstance(s, Atom):
                if s in non_atom_ants: return sigma
            else:
                candidates = index.query(s)
                for a_atom in candidates:
                    term_pairs = list(zip(a_atom.args, s.args))
                    res = unify(sigma, term_pairs, is_substitutable=self._is_metavar)
                    if res is not None:
                        return res
        return None

    def prove(self, formula: Formula) -> Tuple[bool, dict]:
        self.start_time = time.time()
        self.steps = 0
        self.max_depth_reached = 0
        initial_sequent = Sequent(antecedent=[], succedent=[formula])
        failed_cache = set()
        
        for limit in range(1, 31):
            if self._check_timeout(): break
            res = self._search(limit, {}, 0, [], set(), failed_cache, initial_sequent)
            if res:
                elapsed = (time.time() - self.start_time) * 1000.0
                return True, {"steps": self.steps, "max_depth": self.max_depth_reached, "time_ms": elapsed}
        
        return False, {"steps": self.steps, "max_depth": self.max_depth_reached, "time_ms": (time.time() - self.start_time) * 1000.0}

    def _canonical_state(self, sigma: dict, sequent: Sequent) -> int:
        ant_norm = sorted([str(apply_to_formula(sigma, f)) for f in sequent.antecedent])
        suc_norm = sorted([str(apply_to_formula(sigma, f)) for f in sequent.succedent])
        return hash((tuple(ant_norm), tuple(suc_norm)))

    def _search(self, depth_limit: int, sigma: dict, metavars: int, history: List[Sequent], seen_literals: Set[str], failed_cache: Set[Tuple[int, int]], sequent: Sequent) -> Optional[Tuple[dict, int]]:
        self.steps += 1
        # print(f"DEBUG: Search depth={depth_limit} sigma={sigma} seq={sequent}")
        self.max_depth_reached = max(self.max_depth_reached, 30 - depth_limit)

        if self._check_timeout(): return None
        if depth_limit < 0: return None
        
        state_hash = self._canonical_state(sigma, sequent)
        if (state_hash, depth_limit) in failed_cache: return None

        # Regularity Condition: Prevent repetition of identical literals along a branch
        current_literals = set()
        for f in sequent.antecedent:
            if f.is_literal():
                current_literals.add(f"L:{apply_to_formula(sigma, f)}")
        for f in sequent.succedent:
            if f.is_literal():
                current_literals.add(f"R:{apply_to_formula(sigma, f)}")
        
        # We only check for newly added literals to avoid pruning surviving literals
        parent_literals = set()
        if history:
            parent = history[-1]
            for f in parent.antecedent:
                if f.is_literal(): parent_literals.add(f"L:{apply_to_formula(sigma, f)}")
            for f in parent.succedent:
                if f.is_literal(): parent_literals.add(f"R:{apply_to_formula(sigma, f)}")
        
        newly_added = current_literals - parent_literals
        if newly_added & seen_literals:
            return None
        
        new_seen = seen_literals | current_literals

        if any(self._sequent_equal(s, sequent) for s in history): return None

        # 1. Axiom Checks
        if any(isinstance(f, Atom) and f.predicate == "⊤" for f in sequent.succedent) or \
           any(isinstance(f, Atom) and f.predicate == "⊥" for f in sequent.antecedent):
            return sigma, metavars
        
        id_sigma = self._find_identity(sigma, sequent.antecedent, sequent.succedent)
        if id_sigma is not None:
            return id_sigma, metavars
        
        if depth_limit == 0: return None

        # 2. Rule Selection with Complexity Ordering
        inv, br, inst = self._find_all_rules(sequent)
        
        def try_rules(rules_list):
            for rule in rules_list:
                res = self._apply_and_search(depth_limit - 1, sigma, metavars, history + [sequent], new_seen, failed_cache, sequent, rule)
                if res: return res
            return None

        res = None
        if inv:
            res = try_rules(inv)
        else:
            res = try_rules(br + inst)
        
        if res is None:
            failed_cache.add(state_hash)
        return res

    def _find_all_rules(self, sequent: Sequent):
        inv_l, br_l, inst_l = self._collect_left_rules(0, sequent.antecedent)
        inv_r, br_r, inst_r = self._collect_right_rules(0, sequent.succedent)
        
        def get_formula(rule_info):
            rule, i = rule_info
            if rule.endswith("_L"): return sequent.antecedent[i]
            return sequent.succedent[i]

        # Prioritise expansion of the "least complex" formula first
        inv = sorted(inv_l + inv_r, key=lambda r: formula_complexity(get_formula(r)))
        br = sorted(br_l + br_r, key=lambda r: formula_complexity(get_formula(r)))
        inst = sorted(inst_l + inst_r, key=lambda r: formula_complexity(get_formula(r)))
        
        return inv, br, inst

    def _collect_left_rules(self, i: int, formulas: List[Formula]):
        if not formulas: return [], [], []
        inv, br, inst = self._collect_left_rules(i + 1, formulas[1:])
        f = formulas[0]
        if isinstance(f, BinaryOp):
            if f.connective == Connective.AND: return [("AND_L", i)] + inv, br, inst
            if f.connective == Connective.OR: return inv, [("OR_L", i)] + br, inst
            if f.connective == Connective.IMPLIES: return inv, [("IMP_L", i)] + br, inst
            if f.connective == Connective.IFF: return [("IFF_L", i)] + inv, br, inst
        if isinstance(f, Negation): return [("NOT_L", i)] + inv, br, inst
        if isinstance(f, QuantifiedFormula):
            if f.quantifier == Quantifier.EXISTS: return [("EXISTS_L", i)] + inv, br, inst
            if f.quantifier == Quantifier.FORALL: return inv, br, [("FORALL_L", i)] + inst
        return inv, br, inst

    def _collect_right_rules(self, i: int, formulas: List[Formula]):
        if not formulas: return [], [], []
        inv, br, inst = self._collect_right_rules(i + 1, formulas[1:])
        f = formulas[0]
        if isinstance(f, BinaryOp):
            if f.connective == Connective.AND: return inv, [("AND_R", i)] + br, inst
            if f.connective == Connective.OR: return [("OR_R", i)] + inv, br, inst
            if f.connective == Connective.IMPLIES: return [("IMP_R", i)] + inv, br, inst
            if f.connective == Connective.IFF: return inv, [("IFF_R", i)] + br, inst
        if isinstance(f, Negation): return [("NOT_R", i)] + inv, br, inst
        if isinstance(f, QuantifiedFormula):
            if f.quantifier == Quantifier.FORALL: return [("FORALL_R", i)] + inv, br, inst
            if f.quantifier == Quantifier.EXISTS: return inv, br, [("EXISTS_R", i)] + inst
        return inv, br, inst

    def _apply_and_search(self, depth, sigma, metavars, history, seen_literals, failed_cache, sequent, rule_info) -> Optional[Tuple[dict, int]]:
        rule, i = rule_info
        new_ant, new_suc = list(sequent.antecedent), list(sequent.succedent)
        
        if rule == "AND_L":
            f = new_ant.pop(i)
            return self._search(depth, sigma, metavars, history, seen_literals, failed_cache, Sequent(new_ant + [f.left, f.right], new_suc))
        if rule == "OR_R":
            f = new_suc.pop(i)
            return self._search(depth, sigma, metavars, history, seen_literals, failed_cache, Sequent(new_ant, new_suc + [f.left, f.right]))
        if rule == "IMP_R":
            f = new_suc.pop(i)
            return self._search(depth, sigma, metavars, history, seen_literals, failed_cache, Sequent(new_ant + [f.left], new_suc + [f.right]))
        if rule == "NOT_L":
            f = new_ant.pop(i)
            return self._search(depth, sigma, metavars, history, seen_literals, failed_cache, Sequent(new_ant, new_suc + [f.formula]))
        if rule == "NOT_R":
            f = new_suc.pop(i)
            return self._search(depth, sigma, metavars, history, seen_literals, failed_cache, Sequent(new_ant + [f.formula], new_suc))
        if rule == "IFF_L":
            f = new_ant.pop(i)
            imp1 = BinaryOp(Connective.IMPLIES, f.left, f.right)
            imp2 = BinaryOp(Connective.IMPLIES, f.right, f.left)
            return self._search(depth, sigma, metavars, history, seen_literals, failed_cache, Sequent(new_ant + [imp1, imp2], new_suc))
        if rule == "IFF_R":
            f = new_suc.pop(i)
            imp1 = BinaryOp(Connective.IMPLIES, f.left, f.right)
            imp2 = BinaryOp(Connective.IMPLIES, f.right, f.left)
            res = self._search(depth, sigma, metavars, history, seen_literals, failed_cache, Sequent(new_ant, new_suc + [imp1]))
            if res: return self._search(depth, res[0], res[1], history, seen_literals, failed_cache, Sequent(new_ant, new_suc + [imp2]))
            return None
        if rule == "EXISTS_L" or rule == "FORALL_R":
            f = (new_ant if rule == "EXISTS_L" else new_suc).pop(i)
            avoid = []
            for fm in sequent.antecedent + sequent.succedent:
                avoid.extend(free_vars(fm))
            x_prime = variant(f.variable.name, avoid)
            body_prime = f.instantiate(Variable(x_prime))
            if rule == "EXISTS_L": return self._search(depth, sigma, metavars, history, seen_literals, failed_cache, Sequent(new_ant + [body_prime], new_suc))
            else: return self._search(depth, sigma, metavars, history, seen_literals, failed_cache, Sequent(new_ant, new_suc + [body_prime]))
        if rule == "AND_R":
            f = new_suc.pop(i)
            res = self._search(depth, sigma, metavars, history, seen_literals, failed_cache, Sequent(new_ant, new_suc + [f.left]))
            if res: return self._search(depth, res[0], res[1], history, seen_literals, failed_cache, Sequent(new_ant, new_suc + [f.right]))
            return None
        if rule == "OR_L":
            f = new_ant.pop(i)
            res = self._search(depth, sigma, metavars, history, seen_literals, failed_cache, Sequent(new_ant + [f.left], new_suc))
            if res: return self._search(depth, res[0], res[1], history, seen_literals, failed_cache, Sequent(new_ant + [f.right], new_suc))
            return None
        if rule == "IMP_L":
            f = new_ant.pop(i)
            res = self._search(depth, sigma, metavars, history, seen_literals, failed_cache, Sequent(new_ant, new_suc + [f.left]))
            if res: return self._search(depth, res[0], res[1], history, seen_literals, failed_cache, Sequent(new_ant + [f.right], new_suc))
            return None
        if rule == "FORALL_L" or rule == "EXISTS_R":
            f = new_ant[i] if rule == "FORALL_L" else new_suc[i]
            mv, next_mv = self._fresh_metavar(metavars)
            body_prime = f.instantiate(Variable(mv))
            if rule == "FORALL_L":
                new_ant = new_ant[:i] + new_ant[i+1:] + [f]
                return self._search(depth, sigma, next_mv, history, seen_literals, failed_cache, Sequent(new_ant + [body_prime], new_suc))
            else:
                new_suc = new_suc[:i] + new_suc[i+1:] + [f]
                return self._search(depth, sigma, next_mv, history, seen_literals, failed_cache, Sequent(new_ant, new_suc + [body_prime]))
        return None
